### PR TITLE
add advanced trait tutorial

### LINKED ISSUE #893: [Feature]How to create multiple resource objects with trait
How to create multiple resource objects with trait, Thanks.

Workload Defintion:
--------
```
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: gateway
  annotations:
    definition.oam.dev/description: "Gateway workload"
spec:
  definitionRef:
    name: gateways.test.com
  extension:
    template: |
      output: {
        apiVersion: "test.com/v1"
        kind:       "Gateway"
        spec: {
          nodes: parameter.nodes
          ports: parameter.ports
        }
      }
      
      parameter: {
        nodes: [string]
        ports: [int]
      }
```

Service Trait Definition:
--------
```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  annotations:
    definition.oam.dev/description: "service the app"
  name: innerservice
spec:
  appliesToWorkloads:
    - gateway
  definitionRef:
    name: "services.test.com" 
  extension:
    template: |-
      output: {
        apiVersion: "v1"
        kind: "List"
        metadata: {
          resourceVersion: ""
          selfLink: ""
        }
        items: [
          for _, s in parameter.services {
            apiVersion: "test.com/v1"
            kind: "Service"
            metadata:{ 
              name: s.name
            }
            spec: {
              serviceName: s.serviceName
              serviceNamespace: s.serviceNamespace 
              servicePorts: [
                for _, p in s.ports {
                  port: p
                  type: "HTTP"
                }
              ]
            }
          }
        ]
      }
      parameter: {
        services: [...{
          name: string
          serviceName: string
          serviceNamespace: string
          ports: [int]
        }]
      }
```

Route Trait Definition:
--------
```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: gatewayrouter
  annotations:
    definition.oam.dev/description: "Add a route for a service"
spec:
  appliesToWorkloads:
    - "gateway"
  workloadRefPath: spec.workloadRef
  definitionRef:
    name: routers.test.com
  extension:
    template: |-
      output: {
        apiVersion: "test.com/v1"
        kind:       "Router"
        spec: {
          gatewayServiceName: context.name 
          gatewayNamespace: parameter.gatewayNamespace
          host: parameter.host
          port: parameter.port
          routers: [
            for _, r in parameter.routers {
              match: {
                url: { 
                  type: "Prefix"
                  value: r.url
                }
              }
              weightedServices: [
                for _, w in r.weightedServices {
                  innerService: {
                    serviceName: w.serviceName
                    serviceNamespace: w.serviceNamespace
                    servicePort: w.servicePort
                  }
                  serviceWeight: w.serviceWeight                
                }
              ]
            }
          ]
        }
      }
      parameter: {
        gatewayNamespace: string
        host: *"*" | string
        port: int
        routers: [...{
          url: *"/" | string
          weightedServices: [...{
            serviceName: string
            serviceNamespace: string
            servicePort: int
            serviceWeight: int
          }]
        }]
      }
```
Expected a trait can produce multiple services.

### PR DESCRIPTION
**Pull request by wonderflow**
fix #893

Introduce some advanced trait feature:

1. Trait generate multiple resources
2. patch trait

More general traits were introduced in this PR.

### CODE REVIEW

**Review by resouer — CHANGES_REQUESTED** [review]
Let's hold this PR a bit. We need a better way to tell this story - it's one of the core features of the whole platform. I will follow up in a separate comment.

**Comment by resouer** [comment]
Here's what I recommend for the doc side: https://github.com/oam-dev/kubevela/issues/922

Specifically for this PR:

1. We need to upgrade Web Service to `Deployment` + `Service` and use it directly as the example of composition workload type. Composition is killer feature for workload type template.
2. `Ingress` + `Service` is not a great example for composition traits, we need to find sth else which **does not include endpoint resources**.
3. HPA is always a great example for trait.
4. For patch style trait, node-affinity (or PVC) is better than replicas; sidecar is good. Always think about the real world use case for this part, not "separation of concerns"

File: docs/en/platform-engineers/advanced-cue.md [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+Apply this newly defined TraitDefinition into our system:
+
+```shell script
+kubectl apply -f https://raw.githubusercontent.com/oam-dev/kubevela/master/docs/examples/advanced-cue/combo.yaml
+```
+
+Then the usage in is as usual in appfile, you need to use `vela traits` to refresh local capabilities and then use `vela up` to run.
+
+Here we use our server-side API `Application Object` for convenience, you don't rely on `vela` command line tool for this.
```
Comment (resouer) [inline docs/en/platform-engineers/advanced-cue.md:77]: Don't do this.

There's no Appfile or vela CLI/dashboard in platform team doc. There's only k8s resources and kubectl.

Similarly, there's no k8s resources and kubectl in the end user doc, only Appfile and Vela interfaces make sense there.

**Comment by wonderflow** [comment]
@resouer Currently we haven't support Composition on workload side.

**Comment by wonderflow** [comment]
I think we can merge this one first and refactor with better story

**Comment by resouer** [comment]
@wonderflow For this PR, please make it align with the structure in https://github.com/oam-dev/kubevela/issues/922.

You can keep the Ingress + Service examples as I can't find out another built-in resource needs composition (which also proves composition is more common on workload side).

File: docs/examples/advanced-cue/app1.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+          - node
+          - server.js
+        image: oamdev/testapp:v1
+        port: 8080
+      traits:
+        - name: new-route
+          properties:
+            domain: test.my.domain
+            http:
+              myservice: 8080
```
Comment (captainroy-hy) [inline docs/examples/advanced-cue/app1.yaml:21]: `/myservice: 8080` 
Create Ingress failed. HTTPIngressPath.Path must begin with a '/'.

File: docs/en/platform-engineers/advanced-cue.md (hunk 1 above)
Comment (wonderflow) [reply docs/en/platform-engineers/advanced-cue.md:77]: deleted

File: docs/examples/advanced-cue/app1.yaml (hunk 2 above)
Comment (wonderflow) [reply docs/examples/advanced-cue/app1.yaml:21]: yes, it's wrong, will fix it

File: docs/examples/advanced-cue/app1.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+          - server.js
+        image: oamdev/testapp:v1
+        port: 8080
+      traits:
+        - name: new-route
+          properties:
+            domain: test.my.domain
+            http:
+              "/api": 8080
+      type: webservice
```
Comment (resouer) [inline docs/examples/advanced-cue/app1.yaml:20]: why `type` comes here?

File: docs/examples/advanced-cue/newroute.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: new-route
```
Comment (resouer) [inline docs/examples/advanced-cue/newroute.yaml:4]: This is exactly `Ingress` trait.

File: docs/examples/advanced-cue/affinity.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "affinity specify node affinity and toleration"
+  name: affinity
```
Comment (resouer) [inline docs/examples/advanced-cue/affinity.yaml:6]: It's `node-affinity`

File: docs/examples/advanced-cue/dynamic-sa.yaml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "dynamically specify service account"
+  name: serviceacc
```
Comment (resouer) [inline docs/examples/advanced-cue/dynamic-sa.yaml:6]: `service-account`, don't use confusing name

File: docs/examples/advanced-cue/for-loop.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: my-svc
```
Comment (resouer) [inline docs/examples/advanced-cue/for-loop.yaml:4]: `expose`

File: docs/examples/advanced-cue/podenv.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "add env into your pods"
+  name: podenv
```
Comment (resouer) [inline docs/examples/advanced-cue/podenv.yaml:6]: `env` is fine, it's inside the application. For deployment environment, we should use `environment`.

File: docs/examples/advanced-cue/kautoscale.yaml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "Specify auto scale by annotation"
+  name: kautoscale
+spec:
+  appliesToWorkloads:
+    - webservice
```
Comment (resouer) [inline docs/examples/advanced-cue/kautoscale.yaml:9]: It only works with knative workload type

File: docs/examples/advanced-cue/app1.yaml (hunk 3 above)
Comment (wonderflow) [reply docs/examples/advanced-cue/app1.yaml:20]: Yes， it's required value in Application object.

File: docs/examples/advanced-cue/newroute.yaml (hunk 4 above)
Comment (wonderflow) [reply docs/examples/advanced-cue/newroute.yaml:4]: yes

File: docs/examples/advanced-cue/kautoscale.yaml (hunk 9 above)
Comment (wonderflow) [reply docs/examples/advanced-cue/kautoscale.yaml:9]: we don't have knative in real example, I add an annotation

File: docs/en/platform-engineers/advanced-cue.md [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+
+Below is an example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: Trait
+metadata:
+  name: auth-service
+spec:
+  template: |
```
Comment (captainroy-hy) [inline docs/en/platform-engineers/advanced-cue.md:338]: this should be a TraitDefinition, right?

Comment (captainroy-hy) [reply docs/en/platform-engineers/advanced-cue.md:338]: maybe [this test case](https://github.com/oam-dev/kubevela/blob/9173ff828f289f0f4a4ea5b580d4b84aaf87c818/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go#L830) can help here.

Comment (wonderflow) [reply docs/en/platform-engineers/advanced-cue.md:338]: yes, that's a mistake, good catch!

File: docs/examples/advanced-cue/app1.yaml (hunk 3 above)
Comment (resouer) [reply docs/examples/advanced-cue/app1.yaml:20]: I mean why this field comes to the end?

**Review by resouer — APPROVED** [review]

**Pull request merged**